# Capstone 2: Data Wrangling

In [1]:
import pandas as pd
import numpy as np

### COVID dataset:

In [2]:
train = pd.read_csv('train.csv')

In [3]:
train.head()

Id Province/State Country/Region   Lat  Long        Date  ConfirmedCases  \
0   1            NaN    Afghanistan  33.0  65.0  2020-01-22             0.0   
1   2            NaN    Afghanistan  33.0  65.0  2020-01-23             0.0   
2   3            NaN    Afghanistan  33.0  65.0  2020-01-24             0.0   
3   4            NaN    Afghanistan  33.0  65.0  2020-01-25             0.0   
4   5            NaN    Afghanistan  33.0  65.0  2020-01-26             0.0   

   Fatalities  
0         0.0  
1         0.0  
2         0.0  
3         0.0  
4         0.0

In [4]:
#How is 'United States' written in this DataFrame? US? USA? United States of America?
train['Country/Region'].unique()

array(['Afghanistan', 'Albania', 'Algeria', 'Andorra',
       'Antigua and Barbuda', 'Argentina', 'Armenia', 'Aruba',
       'Australia', 'Austria', 'Azerbaijan', 'Bahrain', 'Bangladesh',
       'Barbados', 'Belarus', 'Belgium', 'Benin', 'Bhutan', 'Bolivia',
       'Bosnia and Herzegovina', 'Brazil', 'Brunei', 'Bulgaria',
       'Burkina Faso', 'Cambodia', 'Cameroon', 'Canada',
       'Central African Republic', 'Chile', 'China', 'Colombia',
       'Congo (Brazzaville)', 'Congo (Kinshasa)', 'Costa Rica',
       "Cote d'Ivoire", 'Croatia', 'Cruise Ship', 'Cuba', 'Cyprus',
       'Czechia', 'Denmark', 'Djibouti', 'Dominican Republic', 'Ecuador',
       'Egypt', 'Equatorial Guinea', 'Estonia', 'Eswatini', 'Ethiopia',
       'Finland', 'France', 'French Guiana', 'Gabon', 'Gambia, The',
       'Georgia', 'Germany', 'Ghana', 'Greece', 'Greenland', 'Guadeloupe',
       'Guam', 'Guatemala', 'Guernsey', 'Guinea', 'Guyana', 'Holy See',
       'Honduras', 'Hungary', 'Iceland', 'India', 'Indonesia

In [5]:
#Isolate rows where 'Country/Region' == 'US'
#reset index
#What does the new df look like?
US_tr= train[train['Country/Region'] == 'US'].reset_index()
print(US_tr.shape)
US_tr.head()

(3654, 9)


index     Id Province/State Country/Region      Lat     Long        Date  \
0  13482  19903        Alabama             US  32.3182 -86.9023  2020-01-22   
1  13483  19904        Alabama             US  32.3182 -86.9023  2020-01-23   
2  13484  19905        Alabama             US  32.3182 -86.9023  2020-01-24   
3  13485  19906        Alabama             US  32.3182 -86.9023  2020-01-25   
4  13486  19907        Alabama             US  32.3182 -86.9023  2020-01-26   

   ConfirmedCases  Fatalities  
0             0.0         0.0  
1             0.0         0.0  
2             0.0         0.0  
3             0.0         0.0  
4             0.0         0.0

In [6]:
#how many missing values per column?
US_tr.isna().sum()

index             0
Id                0
Province/State    0
Country/Region    0
Lat               0
Long              0
Date              0
ConfirmedCases    0
Fatalities        0
dtype: int64

In [7]:
#Find max date in COVID dataset to determine date range to limit temp dataset to:
max(US_tr['Date'])

'2020-03-24'

In [8]:
min(US_tr['Date'])

'2020-01-22'

In [9]:
#Remove columns we don't need:
US_tr = US_tr.drop(columns=['index', 'Id', 'Country/Region', 'Lat', 'Long'])

In [10]:
#Rename Province/State column to 'State'
US_tr.rename(columns = {'Province/State':'State'}, inplace = True)

In [11]:
#Is Date column correct dtype?
US_tr['Date'].dtypes

dtype('O')

In [12]:
#Update Date column to datetime object:
US_tr['Date']=pd.to_datetime(US_tr['Date'])

In [13]:
US_tr['Date'].dtypes

dtype('<M8[ns]')

In [14]:
listUS_tr = US_tr['State'].unique()

In [15]:
len(listUS_tr)

58

In [16]:
listUS_tr

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'Delaware', 'Diamond Princess',
       'District of Columbia', 'Florida', 'Georgia', 'Grand Princess',
       'Guam', 'Hawaii', 'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas',
       'Kentucky', 'Louisiana', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Mississippi', 'Missouri', 'Montana',
       'Nebraska', 'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico',
       'New York', 'North Carolina', 'North Dakota', 'Ohio', 'Oklahoma',
       'Oregon', 'Pennsylvania', 'Puerto Rico', 'Rhode Island',
       'South Carolina', 'South Dakota', 'Tennessee', 'Texas', 'US',
       'United States Virgin Islands', 'Utah', 'Vermont',
       'Virgin Islands', 'Virginia', 'Washington', 'West Virginia',
       'Wisconsin', 'Wyoming'], dtype=object)

In [17]:
US_tr[US_tr['State']=='US']

State       Date  ConfirmedCases  Fatalities
3024    US 2020-01-22             0.0         0.0
3025    US 2020-01-23             0.0         0.0
3026    US 2020-01-24             0.0         0.0
3027    US 2020-01-25             0.0         0.0
3028    US 2020-01-26             0.0         0.0
...    ...        ...             ...         ...
3082    US 2020-03-20             0.0         0.0
3083    US 2020-03-21             0.0         0.0
3084    US 2020-03-22             0.0         0.0
3085    US 2020-03-23             0.0         0.0
3086    US 2020-03-24             0.0         0.0

[63 rows x 4 columns]

In [18]:
US_tr= US_tr[US_tr['State'] != 'US']

In [19]:
len(US_tr['State'].unique())

57

### Temperature Dataset:

In [20]:
#Import COVID dataset:
temp_df = pd.read_csv('city_temperature.csv')

/Users/abigailmorgan/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3165: DtypeWarning: Columns (2) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [21]:
temp_df.head()

Region  Country State     City  Month  Day  Year  AvgTemperature
0  Africa  Algeria   NaN  Algiers      1    1  1995            64.2
1  Africa  Algeria   NaN  Algiers      1    2  1995            49.4
2  Africa  Algeria   NaN  Algiers      1    3  1995            48.8
3  Africa  Algeria   NaN  Algiers      1    4  1995            46.4
4  Africa  Algeria   NaN  Algiers      1    5  1995            47.9

In [22]:
#How is 'United States' written in this DataFrame? US? USA? United States of America?
temp_df['Country'].unique()

array(['Algeria', 'Burundi', 'Benin', 'Central African Republic', 'Congo',
       'Egypt', 'Ethiopia', 'Gabon', 'Gambia', 'Guinea', 'Guinea-Bissau',
       'Ivory Coast', 'Kenya', 'Morocco', 'Madagascar', 'Mauritania',
       'Malawi', 'Mozambique', 'Namibia', 'Nigeria', 'Senegal',
       'Sierra Leone', 'South Africa', 'Togo', 'Tunisia', 'Tanzania',
       'Uganda', 'Zambia', 'Bangladesh', 'China', 'Hong Kong', 'India',
       'Indonesia', 'Japan', 'Kazakhstan', 'Kyrgyzstan', 'Laos',
       'Malaysia', 'Mongolia', 'Myanmar (Burma)', 'Nepal', 'North Korea',
       'Pakistan', 'Philippines', 'Singapore', 'South Korea', 'Sri Lanka',
       'Taiwan', 'Tajikistan', 'Thailand', 'Turkmenistan', 'Uzbekistan',
       'Vietnam', 'Australia', 'New Zealand', 'Albania', 'Austria',
       'Belarus', 'Belgium', 'Bulgaria', 'Croatia', 'Cyprus',
       'Czech Republic', 'Denmark', 'Finland', 'France', 'Germany',
       'Georgia', 'Greece', 'Hungary', 'Iceland', 'Ireland', 'Italy',
       'Latvia', 'Ma

In [23]:
#Isolate rows where 'Country/Region' == 'US'
#reset index
#What does the new df look like?
US_temp= temp_df[temp_df['Country'] == 'US'].reset_index()
print(US_temp.shape)
US_temp.head()

(1455337, 9)


index         Region Country    State        City  Month  Day  Year  \
0  1450990  North America      US  Alabama  Birmingham      1    1  1995   
1  1450991  North America      US  Alabama  Birmingham      1    2  1995   
2  1450992  North America      US  Alabama  Birmingham      1    3  1995   
3  1450993  North America      US  Alabama  Birmingham      1    4  1995   
4  1450994  North America      US  Alabama  Birmingham      1    5  1995   

   AvgTemperature  
0            50.7  
1            37.2  
2            33.2  
3            33.3  
4            26.4

In [24]:
#how many missing values per column?
US_temp.isna().sum()

index             0
Region            0
Country           0
State             0
City              0
Month             0
Day               0
Year              0
AvgTemperature    0
dtype: int64

In [25]:
#Datatypes?
US_temp.dtypes

index               int64
Region             object
Country            object
State              object
City               object
Month               int64
Day                 int64
Year                int64
AvgTemperature    float64
dtype: object

In [26]:
#Combine month, year, day columns to single date column  
cols=["Year","Month","Day"]
US_temp['Date'] = US_temp[cols].apply(lambda x: '-'.join(x.values.astype(str)), axis="columns")

In [27]:
#Convert to datetime object
US_temp['Date']=pd.to_datetime(US_temp['Date'])

In [28]:
#Check dtype
US_temp['Date'].dtypes

dtype('<M8[ns]')

In [29]:
#Remove columns we don't need
US_temp = US_temp.drop(columns=['Region', 'Country', 'City', 'Month', 'Day', 'Year', 'index'])

In [30]:
#Specify date range matching COVID dataset (01/22/2020- 03/24/2020)
US_temp_range = US_temp.loc[(US_temp['Date']>= '2020-01-22') & (US_temp['Date']<= '2020-03-24')].reset_index()

In [31]:
#Remove US rows we don't need (Additional Territories)
US_temp_range = US_temp_range[US_temp_range['State'] != 'Additional Territories']

In [33]:
states_list = US_temp_range['State'].unique()

In [34]:
len(states_list)

50

In [35]:
states_list

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'District of Columbia', 'Maryland',
       'Florida', 'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana',
       'Iowa', 'Kansas', 'Kentucky', 'Louisiana', 'Maine',
       'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi',
       'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina',
       'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
       'West Virginia', 'Wisconsin', 'Wyoming'], dtype=object)

### Remove remaining non-matching 'States'

In [49]:
US_cov = US_tr[US_tr['State'] != 'Delaware']
US_cov = US_cov[US_cov['State'] != 'Diamond Princess']
US_cov = US_cov[US_cov['State'] != 'Grand Princess']
US_cov = US_cov[US_cov['State'] != 'Guam']
US_cov = US_cov[US_cov['State'] != 'Puerto Rico']
US_cov = US_cov[US_cov['State'] != 'United States Virgin Islands']
US_cov = US_cov[US_cov['State'] != 'Virgin Islands']

In [51]:
US_cov['State'].unique()

array(['Alabama', 'Alaska', 'Arizona', 'Arkansas', 'California',
       'Colorado', 'Connecticut', 'District of Columbia', 'Florida',
       'Georgia', 'Hawaii', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Kansas', 'Kentucky', 'Louisiana', 'Maine', 'Maryland',
       'Massachusetts', 'Michigan', 'Minnesota', 'Mississippi',
       'Missouri', 'Montana', 'Nebraska', 'Nevada', 'New Hampshire',
       'New Jersey', 'New Mexico', 'New York', 'North Carolina',
       'North Dakota', 'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Texas', 'Utah', 'Vermont', 'Virginia', 'Washington',
       'West Virginia', 'Wisconsin', 'Wyoming'], dtype=object)

In [52]:
len(US_cov['State'].unique())

50

### Merge to one DataFrame:

In [53]:
dfnew = pd.merge(US_tr, US_temp_range, how = 'left', left_on= ['State', 'Date'], right_on= ['State', 'Date'])

In [54]:
dfnew.head()


State       Date  ConfirmedCases  Fatalities    index  AvgTemperature
0  Alabama 2020-01-22             0.0         0.0   9152.0            35.4
1  Alabama 2020-01-22             0.0         0.0  18417.0            32.0
2  Alabama 2020-01-22             0.0         0.0  27682.0            36.1
3  Alabama 2020-01-22             0.0         0.0  36947.0            36.0
4  Alabama 2020-01-23             0.0         0.0   9153.0            43.0

In [55]:
#Remove Extra index column:
dfnew = dfnew.drop(columns='index')

In [56]:
#Confirm all dtypes are appropriate
dfnew.dtypes

State                     object
Date              datetime64[ns]
ConfirmedCases           float64
Fatalities               float64
AvgTemperature           float64
dtype: object

In [57]:
#Any null values?
dfnew.isnull().sum()

State               0
Date                0
ConfirmedCases      0
Fatalities          0
AvgTemperature    441
dtype: int64

## Summary:

The original COVID-19 dataset was first abbreviated to only include data from the United States. The resulting dataframe was 3654 row and 9 columns. A preliminary check on Null values revealed this dataset had none. Next, in order to determine the date range of the information contained in the COVID dataset, the maximum (latest) date from the DataFrame was determined to be 2020-03-24. The startdate could be determined from viewing the first few rows of the DataFrame, with df.head(), but this starting date was confirmed by searching for the minimum date to be: 2020-01-22. This information would be used to subset the second temperature DataFrame, which contained daily temperature values from 1995 until the present day.

Next, irrelevant columns were removed from the COVID-19 dataset, including: 'index', 'Id', 'Country/Region', 'Lat', 'Long'. Thus, five columns in total were removed. The name of the 'Province/State' was renamed to 'State', in order to match the column name of the second Dataframe that would be merged to it in later steps. The data type of 'Date' column was changed from object to datetime object. Lastly, a list was made of all the unique values of the 'States' column to determine if the Temperature dataset include the same list of 'States' in later steps. This list of unique states had a length of 58 (8 higher than the number of states in the US), and printing the list revealed that, in addition to the traditional 50 states, the COVID-19 dataset also provided information on 8 other regions marked as US territories (but not states):
1. District of Columbia
2. Diamond Princess
3. Grand Princess
4. Guam
5. Puerto Rico
6. US
7. United States Virgin Islands
8. Virgin Islands

The 'State' listed as 'US' was further examined to see if it contained any relevant information. There were 63 rows of data for the 'State' of 'US', all of which were empty. These rows were removed, as they didn't contain any data and 'US' is not a 'State' anyways.

For now, the rest of the US territories are left in the COVID-19 dataset, as they do contain data. If the temperatures dataset has data for any of those regions as well, it may be worthwhile to include. There are now 57 unique values for 'State' in the COVID-19 DataFrame (50 states and seven territories).

The second dataset brought in was a Temperature dataset containing daily average temperatures in major cities across the globe. This was a very large dataset with far more information that was needed, so initial steps were to trim down the excess information. First, the DataFrame was reduced to only data pertaining the the United States. The resulting DataFrame had //1,455,337 rows , and 9 columns, and contained no Null values. In this temperature DataFrame, the dates for each record were reported in three separate columns: Month, Day, and Year, all of which were integer64. These three values were combined to create a separate 'Date' column in the format of YYYY-MM-DD to match the COVID-19 dataset. The column was then converted to datetime object type to match the 'Date' column type from the COVID-19 dataset. The irrelevant columns ('Region', 'Country', 'City', 'Month', 'Day', 'Year', 'index') from the temperature DataFrame were also dropped. A total of six columns were dropped from the Temperature DataFrame and one column was added ('Date'). The temperature DataFrame was further shorted by limiting the date range to the date range of available data in the COVID-19 DataFrame (which was previously determined to be 2020-01-22 to 2020-03-42). All rows containing temperature data outside of this timeframe were removed. At the tail of the DataFrame, values for "Additional US territories" were recorded, but without any further information on which territories they were, it would be impossible to match them to any COVID-19 data, so these rows were dropped.

Next, the values for 'States' in the temperature DataFrame were examined. This DataFrame contained 50 unique values in the 'States' column, precisely the number of states in the US. Upon further examination of these values, however, 'District of Columbia' was included (which is not a state) and 'Delaware' was excluded. Because there is information in the COVID-19 dataset for the 'District of Columbia', rows containing that value for 'State' were retained. 

Before merging the datasets, each DataFrame needed to contain only information on 'States' included in both data sets. Therefore, as 'Delaware' was somehow missing from the Temperature dataset, rows pertaining to that value in the COVID-19 dataset were also dropped. As for the list of "extra" US Territories previously found in the COVID-19 dataset, 'District of Columbia' was included in the temperature dataset, so rows pertaining to that region were kept. All other territories, however, were removed. Now both DataFrame contained exactly 50 unique values for 'State'.

Next the COVID-19 and Temperature DataFrames were merged together using a left join on both State and Date columns to match daily COVID-19 records with daily average temperatures by state. The index column was dropped as well. Datatypes were checked and appear to all be appropriate types. There are now 441 missing values for Average Temperature, which will need to be further examined in the next steps. 